Restructure smart contract section (#700)
dfinity/portal: how-it-works/7_smart_contracts/01_canister-lifecycle.card.md

---
title: Canisters
---

![](/img/how-it-works/canister-lifecycle.600x300.jpg)

# Canisters

Smart contracts on the Internet computer come in the form of *canisters*: computational units that bundle together code and state. Each canister defines functions that can be called by other canisters and parties external to the IC, such as browsers or mobile apps.
Canisters communicate with one another via asynchronous messages but the execution of each such message is done in complete isolation, allowing for massive levels of concurrent execution. 

> Canisters are managed by controllers. Control structure of canisters could be centralized (e.g. when the controllers include some centralized entity), descentralized (when the controller is a DAO) or even not existent, in which case the canister is an immutable smart contract. 

Author: ```suggestion
Canisters are managed by controllers. Control structure of canisters could be centralized (e.g. when the controllers include some centralized entity), decentralized (when the controller is a DAO) or even non-existent, in which case the canister is an immutable smart contract. 
```

---
title: Canisters
---

![](/img/how-it-works/canister-lifecycle.600x300.jpg)

# Canisters

> On the Internet Computer smart contracts come in the form of canisters.  These are computational units which bundle together code and state. Each canister defines a set of entry points which can be called by other canisters and parties external to the IC, such as browsers or mobile apps.

Author: ```suggestion
Smart contracts on the Internet computer come in the form of *canisters*: computational units that bundle together code and state. Each canister defines functions that can be called by other canisters and parties external to the IC, such as browsers or mobile apps.
```

> When deployed, canisters behave much like actors from the actor-based concurrency model – they communicate asynchronously with each other but, unlike in the traditional actor model, communication is bidirectional. 

Author: I think most readers will not have heard of the actor model – so it may be preferable to formulate this part without referring to that model. I like the way you use and introduce it in the "Go Deeper" section, though. But it would not seem appropriate to do the same here.

Reviewer: Removed

> Canisters are managed by controllers. These can deploy the canister to the Internet Computer, start/stop their execution and update their code.  As they execute, canisters need to acquire resources for their execution (memory, network bandwidth and computational power). To this end canisters hold, locally, a cycle balance which is the unit used to measure resource consumption. The controllers also need to ensure that canisters hold sufficient cycles. 

Author: Overall, this seems good. I was just thinking whether it makes sense to more immediately say that controllers can be any of (a) centralized, (b) empty – which makes the canister immutable – or (c) governed by a DAO. I think the notion of a *controller* may evoke images of centralized control, which is possible but not necessary. Anyway, just my reading and I may be wrong.

Reviewer: Implemented
